### PR TITLE
adds HLTPixelActivityHFSumEnergyFilter module (superseeds #12483)

### PR DESCRIPTION
**Pull request by mverwe**
This PR superseeds #12483

Adds HLTPixelActivityHFSumEnergyFilter module. This filter checks correlation between number of pixel clusters and HF energy sum. The idea is to have a filter firing when the expected correlation is broken.

Remark 1 : This module is NOT for the pp reference run.
Remark 2 : This module is for PbPb run.

@kkrajczar 
@icali
@ttrk

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@mverwe please use "Edit" button at the top and change the title of the PR to something meaningful and self-descriptive

**Comment by Martin-Grunewald** [comment]
This must be a normal (stream or global) EDFilter, not an HLTFilter, as it does not (and can not!) save any TriggerObjects. Please make the change.

**Comment by Martin-Grunewald** [comment]
-1

**Comment by Martin-Grunewald** [comment]
please test

File: HLTrigger/special/src/HLTPixelActivityHFSumEnergyFilter.cc [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    if (it->energy()>eCut_HF_) {
+      sumE += it->energy();
+    }
+  }
+
+  bool accept = kFALSE;
+
+  double thres = offset_ + slope_ * sumE;
+  double diff = clusterSize - thres;
+  if(sumE>eMin_HF_ && diff < maxDiff_) accept = kTRUE;
```
Comment (Martin-Grunewald) [inline HLTrigger/special/src/HLTPixelActivityHFSumEnergyFilter.cc]: Should'nt that be std::abs(diff)?

File: HLTrigger/special/src/HLTPixelActivityHFSumEnergyFilter.cc [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  double sumE = 0.;
+
+  for (HFRecHitCollection::const_iterator it=HFRecHitsH->begin(); it!=HFRecHitsH->end(); it++) {
+    if (it->energy()>eCut_HF_) {
+      sumE += it->energy();
+    }
+  }
+
+  bool accept = kFALSE;
```
Comment (Martin-Grunewald) [inline HLTrigger/special/src/HLTPixelActivityHFSumEnergyFilter.cc]: kTrue and kFalse is ROOT jargon, please use C++ true/false.

**Comment by mverwe** [comment]
@Martin-Grunewald 
The trigger should fire if the measured value is smaller than the expected value and not if it is larger. See slide 2 here: https://indico.cern.ch/event/460253/contribution/8/attachments/1190230/1727697/151119_HFSumEnergyFilter.pdf

**Comment by Martin-Grunewald** [comment]
Still, currently you accept events with diff=0 which seems wrong even in slide 2.

**Comment by mverwe** [comment]
I don't understand the objection: we want to trigger on the events in lower right triangle. We will tune the parameters of the line based on data and place it well below the expected correlation. maxDiff_ will be zero or a small value if that is better for numerical reasons which I thought is what you meant in the comments of the previous PR.

**Comment by Martin-Grunewald** [comment]
Events in the lower right triangle do not have diff=0. The filter code accepts events with diff=0.
diff =0 corresponds to the diagonal line (at 45 degrees) with perfect agreement between the two numbers plotted.

**Comment by mverwe** [comment]
If we set maxDiff_ to zero, events with diff==0 will not be accepted. Are you asking me to set the default value of maxDiff_ to zero? or to remove it and change the line to:
if(sumE>eMin_HF_ && diff < 0.) accept = kTRUE;

**Comment by Martin-Grunewald** [comment]
The code setts maxDiff to:
```
desc.add<double>("maxDiff",1e-5);
```
which is not zero but positive.

Anyway, the claim was that events with NON-perfect correlation should be selected.
NON-perfect means the two values differ. Thus you should not accpet diff=0
events. 
f course, if the description of the logic of the filter in the comment section is wrong,
please correct it in the comment. If the C++ condition is wrong, fix that. 
Currently both seem incompatible!

File: HLTrigger/special/src/HLTPixelActivityHFSumEnergyFilter.cc [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+void
+HLTPixelActivityHFSumEnergyFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("inputTag",edm::InputTag("hltSiPixelClusters"));
+  desc.add<edm::InputTag>("HFHitCollection",edm::InputTag("hltHfreco"));
+  desc.add<double>("eCut_HF",0);
+  desc.add<double>("eMin_HF",0);
+  desc.add<double>("offset",0);
+  desc.add<double>("slope",0);
```
Comment (Martin-Grunewald) [inline HLTrigger/special/src/HLTPixelActivityHFSumEnergyFilter.cc]: Why is slope_ set to zero?

**Comment by mverwe** [comment]
I will remove maxDiff as an option since it only causes confusion. The naming is indeed not optimal.
The slope and offset parameter are not set yet because we don't have PbPb run2 data yet so we don't know the current correlation. We understood we can set these values from outside in confdb. Is this incorrect? In necessary, we can provide default values based on run1 PbPb data.

**Comment by Martin-Grunewald** [comment]
Anyway, wouldn't you also need to select events in an upper (leftish) triangle away from the blue scatter line?
Or is it otherwise clear that there will be never any events in that region?

**Comment by ttrk** [comment]
Let me try to explain what I understand. Before calculating the "diff" we apply an offset so that it becomes diff = clustersize - (correlation line + offset). So "diff == 0" will correspond to red line on slide 2. The offset shifts the purple correlation line to the red line (slide 3). In short, "diff" is the difference from the red line, not from the correlation line. 
https://indico.cern.ch/event/460253/contribution/8/attachments/1190230/1727697/151119_HFSumEnergyFilter.pdf

**Comment by Martin-Grunewald** [comment]
Ah I see, so you do not expect slope=1 at all. OK, technically this is just a place holder value and 
the real value will indeed be set via the GUI for the specific menu.

**Comment by mverwe** [comment]
The events in the upper triangle are selected by a different filter which also uses a rough primary vertex finder. This was already used during run1 and we don't see the need to also try to select them with this filter which is specifically for when the pixel didn't empty FIFO.
Ok, we will add default parameters based on run1.

**Comment by Martin-Grunewald** [comment]
Ah right, thanks for the explanation of diff!
(Please add it as a comment into the code where diff is calculated,
as I am sure other people will be tripped up as well :)

**Comment by mverwe** [comment]
We've addressed all comments. Let us know in case of further remarks.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by slava77** [comment]
merge

all new code

**Comment by Martin-Grunewald** [comment]
BTW, is there a PR for 76X as well? Please make one!

**Comment by mverwe** [comment]
@Martin-Grunewald The PR for 76X will be done asap

**Pull request merged**